source/validate.h

```diff
@@ -84,91 +103,293 @@ enum class FunctionDecl {
+
+  /// Registers the end of the block
+  void RegisterBlockEnd(SpvOp branch_instruction);
+
+  /// Registers the end of the block
+  void RegisterBlockEnd(uint32_t next_id, SpvOp branch_instruction);
+
+  /// Registers the end of the block
+  void RegisterBlockEnd(std::vector<uint32_t> next_list, SpvOp branch_instruction);
+
+  /// Returns true if the \p merge_block_id is a merge block
+  bool IsMergeBlock(uint32_t merge_block_id) const;
+
+  /// Returns true if the \p id is the first block of this function
+  bool IsFirstBlock(uint32_t id) const;
 
-  spv_result_t RegisterBlockSelectionMerge(uint32_t merge_id,
-                                           SpvSelectionControlMask control);
+  /// Returns the first block of the current function
+  const BasicBlock* get_first_block() const;
 
-  // Registers the end of the block
-  spv_result_t RegisterBlockEnd();
+  /// Returns the first block of the current function
+  BasicBlock* get_first_block();
+
+  /// Returns a vector of all the blocks in the function
+  const std::vector<BasicBlock*>& get_blocks() const;
+
+  /// Returns a vector of all the blocks in the function
+  std::vector<BasicBlock*>& get_blocks();
+
+  /// Returns a list of all the cfg constructs in the function
+  const std::list<CFConstruct>& get_constructs() const;
+
+  /// Returns a list of all the cfg constructs in the function
+  std::list<CFConstruct>& get_constructs();
 
   // Returns the number of blocks in the current function being parsed
   size_t get_block_count() const;
 
-  // Returns true if called after a function instruction but before the
-  // function end instruction
-  bool in_function_body() const;
+  /// Returns the id of the funciton
+  uint32_t get_id() const { return id_; };
 
-  // Returns true if called after a label instruction but before a branch
-  // instruction
+  // Returns the number of blocks in the current function being parsed
+  size_t get_undefined_block_count() const;
+  const std::unordered_set<uint32_t>& get_undefined_blocks() const {
+    return undefined_blocks_;
+  }
+
+  /// Returns true if called after a label instruction but before a branch
+  /// instruction
   bool in_block() const;
 
-  libspirv::DiagnosticStream diag(spv_result_t error_code) const;
+  /// Returns the block that is currently being parsed in the binary
+  BasicBlock& get_current_block();
+
+  /// Returns the block that is currently being parsed in the binary
+  const BasicBlock& get_current_block() const;
+
+  /// Prints a GraphViz digraph of the CFG of the current funciton
+  void printDotGraph() const;
+
+  /// Prints a directed graph of the CFG of the current funciton
+  void printBlocks() const;
 
  private:
-  // Parent module
+  /// Parent module
   ValidationState_t& module_;
 
-  // Function IDs in a module
-  std::vector<uint32_t> id_;
+  /// Function ID of
+  uint32_t id_;
 
-  // OpTypeFunction IDs of each of the id_ functions
-  std::vector<uint32_t> type_id_;
+  /// The type of the function
```

Comment:
This refers to the OpTypeFunction id for this function. The result_type_id is the OpType* of the return type.